File: learning-factory/warehouse/artifacts/projects/build-your-own-x/programming-language/the-super-tiny-compiler-go-0095b2554e/job_byox_build_s2_0aadb1e25ccffd07eeb644fba4d16b87/attempt-001/sealed/reference/pipeline.go
package prefixforge

import "io"

func Execute(source string, out io.Writer) (Value, error) {
	tokens, err := Tokenize(source)
	if err != nil {
		return Value{}, err
	}
	program, err := Parse(tokens)
	if err != nil {
		return Value{}, err
	}
	code, err := Compile(program)
	if err != nil {
		return Value{}, err
	}
	return Run(code, out)
}

File: learning-factory/warehouse/artifacts/projects/build-your-own-x/programming-language/the-super-tiny-interpreter-27debb2830/job_byox_build_s2_aace1e25a85a7f45f3e8b88474341820/attempt-001/sealed/reference/src/pipeline.mjs
import { compile } from "./compiler.mjs";
import { interpret } from "./interpreter.mjs";
import { tokenize } from "./lexer.mjs";
import { parse } from "./parser.mjs";
import { run } from "./vm.mjs";

export function execute(source, { backend = "tree" } = {}) {
  const program = parse(tokenize(source));
  if (backend === "tree") return interpret(program);
  if (backend === "vm") return run(compile(program));
  throw new TypeError(`unknown backend: ${backend}`);
}
